sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -680,6 +681,14 @@ class Dataset[T] private[sql](
     joinWith(other, condition, "inner")
   }
 
+  /**
+   * Joins this [[Dataset]] returning a [[Tuple2]] for each pair using cartesian join
+   * Note: cartesian joins are very expensive without a filter that can be pushed down.
+   *
+   * @since 2.0.0
+   */
+  def joinWith[U](other: Dataset[U]): Dataset[(T, U)] = joinWith(other, lit(true), "inner")
```

Comment:
Thanks for your feedback @marmbrus . The only join API in Dataset I can find is:

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L644

which expects a Column. Do you mean to add some other method like the one in Dataframe:

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala#L383-L385

If so, I'm wondering whether we need to refactor out the code that handles encoder?